sql: TestReacquireLeaseOnRestart failed

sql.TestReacquireLeaseOnRestart [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16567338?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16567338?buildTab=artifacts#/) on master @ [5c51c69e82c471e6667489a828f191940e5c425e](https://github.com/cockroachdb/cockroach/commits/5c51c69e82c471e6667489a828f191940e5c425e):


```
=== RUN   TestReacquireLeaseOnRestart
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/492cc08314ebb30515b6f60e383090f4/logTestReacquireLeaseOnRestart3185703012
    test_log_scope.go:81: use -show-logs to present logs inline
    txn_restart_test.go:1221: pq: failed to insert lease {[128] 105 1 1 {{18184000 63859861489 <nil>}} [1 1 128 181 251 85 191 210 136 78 120 162 37 167 116 99 9 168 185]}: unexpected value: raw_bytes:"\035HOc\n3\002" timestamp:<wall_time:1724263801038456666 > 
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/492cc08314ebb30515b6f60e383090f4/logTestReacquireLeaseOnRestart3185703012
--- FAIL: TestReacquireLeaseOnRestart (0.53s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReacquireLeaseOnRestart.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-41537

--- a/pkg/sql/catalog/lease/descriptor_state.go
+++ b/pkg/sql/catalog/lease/descriptor_state.go
@@ -172,6 +172,10 @@ func (t *descriptorState) upsertLeaseLocked(
 	}
 	if session != nil {
 		s.mu.lease.sessionID = session.ID().UnsafeBytes()
+		// When using session based leasing, if we end up acquiring the same lease again
+		// nothing needs to be cleaned up or updated. This is because the system.lease
+		// table does not store any expiry inside the table.
+		toRelease.sessionID = nil
 	}
 	if log.ExpensiveLogEnabled(ctx, 2) {
 		log.VEventf(ctx, 2, "replaced lease: %s with %s", toRelease, s.mu.lease)

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -812,16 +812,8 @@ func purgeOldVersions(
 	// Acquire a refcount on the descriptor on the latest version to maintain an
 	// active lease, so that it doesn't get released when removeInactives()
 	// is called below. Release this lease after calling removeInactives().
-	//
-	// If the lease ends up being expired anyway it's okay to purge all previous
-	// versions (assuming no one has a ref count). With the session based
-	// leasing upgrade and tests with zero lease duration, we can have the lease
-	// expire right after acquiring it. Because renewals are disabled in later
-	// stages of this migration there is no other mechanism to purge old versions
-	// if we hit this case. Note: This scenario is impossible to hit in the real
-	// world since the lease duration is never set to 0.
 	desc, _, err := t.findForTimestamp(ctx, m.storage.clock.Now())
-	if isInactive := catalog.HasInactiveDescriptorError(err); err == nil || isInactive || errors.Is(err, errRenewLease) {
+	if isInactive := catalog.HasInactiveDescriptorError(err); err == nil || isInactive {
 		removeInactives(isInactive)
 		if desc != nil {
 			t.release(ctx, desc)

